src/_pytest/assertion/rewrite.py

```diff
@@ -954,7 +954,9 @@ def visit_Call_35(self, call):
         visit `ast.Call` nodes on Python3.5 and after
         """
         if isinstance(call.func, ast.Name) and call.func.id == "all":
```

Comment:
I think we should put the guard here instead (functions returning `Optional[T]` are a bit odd)

there's also another edgecase which is not handled:

```python
if all(*f):
    ...
```

(in python2 this will `IndexError`)

Reply:
Here is how I handle this in `pyupgrade`: https://github.com/asottile/pyupgrade/blob/8ba7b8b1844f76d2f3a8233cb1a632f3358fea9f/pyupgrade.py#L1368-L1378

though I think in this case it's not super necessary to identify it as starargs or not (since in py2 you'll end up with `args=[]` and in py35+ you'll end up with `args=Starred(...)` both of which would be caught by a length and isinstance check

So basically:

```python
if (
        isinstance(call.func, ast.Name) and
        call.func.id == 'all' and
        len(call.args) == 1 and
        isinstance(call.args[0], (...))
):
    return self._visit_call(...)
```

Might even be useful to isolate to a function so you can do

```python
if _is_all_call(call):
    return self._visit_call(call)
```

Reply:
> I think we should put the guard here instead (functions returning Optional[T] are a bit odd)

Good point, thanks. Done.

> if all(*f):

What do you mean? The assertion rewriter only kicks in for `assert` statements, not for `if`s.

Reply:
ah yeah I meant `assert all(*f)`:

```python
def test():
    f = ((1,))
    assert all(*f)
```

```console
$ pytest t.py 
============================= test session starts ==============================
platform linux2 -- Python 2.7.15rc1, pytest-4.6.0, py-1.8.0, pluggy-0.12.0
rootdir: /tmp
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
____________________________ ERROR collecting t.py _____________________________
venv/local/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:169: in find_module
    source_stat, co = _rewrite_test(self.config, fn_pypath)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:404: in _rewrite_test
    rewrite_asserts(tree, fn, config)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:453: in rewrite_asserts
    AssertionRewriter(module_path, config).run(mod)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:701: in run
    new.extend(self.visit(child))
/usr/lib/python2.7/ast.py:241: in visit
    return visitor(node)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:828: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
/usr/lib/python2.7/ast.py:241: in visit
    return visitor(node)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:1013: in visit_Call_legacy
    return self._visit_all(call)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:983: in _visit_all
    if not isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp)):
E   IndexError: list index out of range
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.36 seconds ============================
```

Reply:
Done! 👍